synapse/handlers/auth.py

```diff
@@ -603,6 +603,87 @@ async def _check_auth_dict(
         (canonical_id, callback) = await self.validate_login(user_id, authdict)
         return canonical_id
 
+    async def username_from_identifier(
+        self, identifier: Dict[str, str], password: Optional[str] = None
+    ) -> Optional[str]:
+        """Given a dictionary containing an identifier from a client, extract the
+        possibly unqualified username of the user that it identifies. Does *not*
+        guarantee that the user exists.
+        If this identifier dict contains a threepid, we attempt to ask password
```

Comment:
Please add blank lines between the different sections in docstrings, it makes it much easier to read!

Reply:
I agree with you, not sure why I did that in the first place :thinking: